tests/interlocked_int/interlocked_int.c

```diff
@@ -0,0 +1,855 @@
+Tests_SRS_INTERLOCKED_43_054: [ interlocked_increment_64 shall return the incremented value. ]
+*/
+TEST_FUNCTION(interlocked_increment_64_upper_bound)
+{
+    ///arrange
+    volatile int64_t addend = INT64_MAX - 1;
+
+    ///act
+    int64_t return_val = interlocked_increment_64(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, addend);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, return_val);
+}
+/*
+Tests_SRS_INTERLOCKED_43_023 [ interlocked_increment_64 shall atomically increment (increase by one) the 64-bit variable *addend.]
+Tests_SRS_INTERLOCKED_43_054: [ interlocked_increment_64 shall return the incremented value. ]
+*/
+TEST_FUNCTION(interlocked_increment_64_lower_bound)
+{
+    ///arrange
+    volatile int64_t addend = INT64_MIN;
+
+    ///act
+    int64_t return_val = interlocked_increment_64(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN + 1, addend);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN + 1, return_val);
+
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_024 [ interlocked_or shall perform an atomic bitwise OR operation on the 32-bit integers *destination and value and store the result in destination.]
+Tests_SRS_INTERLOCKED_43_055: [ interlocked_or shall return the initial value of *destination. ]
+*/
+
+TEST_FUNCTION(interlocked_or_does_bitwise_or)
+{
+    ///arrange
+    volatile uint32_t destination = 0xF0F0F0F0;
+    uint32_t value = 0x0F0F0F0F;
+
+    ///act
+    uint32_t return_val = (uint32_t)interlocked_or((volatile int32_t*)&destination, (int32_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(uint32_t, -1, destination);
+    ASSERT_ARE_EQUAL(uint32_t, 0xF0F0F0F0, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_025 [ interlocked_or_16 shall perform an atomic bitwise OR operation on the 16-bit integers *destination and value and store the result in destination.]
+Tests_SRS_INTERLOCKED_43_056: [ interlocked_or_16 shall return the initial value of *destination. ]
+*/
+
+TEST_FUNCTION(interlocked_or_16_does_bitwise_or)
+{
+    ///arrange
+    volatile uint16_t destination = 0xF0F0;
+    uint16_t value = 0x0F0F;
+
+    ///act
+    uint16_t return_val = (uint16_t)interlocked_or_16((volatile int16_t*)&destination, (int16_t)value);
+
+
+    ///assert
+    ASSERT_ARE_EQUAL(uint16_t, -1, destination);
+    ASSERT_ARE_EQUAL(uint16_t, 0xF0F0, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_026 [ interlocked_or_64 shall perform an atomic bitwise OR operation on the 64-bit integers *destination and value and store the result in destination.]
+Tests_SRS_INTERLOCKED_43_057: [ interlocked_or_64 shall return the initial value of *destination. ]
+*/
+
+TEST_FUNCTION(interlocked_or_64_does_bitwise_or)
+{
+    ///arrange
+    volatile uint64_t destination = 0xF0F0F0F0F0F0F0F0;
+    uint64_t value = 0x0F0F0F0F0F0F0F0F;
+
+    ///act
+    uint64_t return_val = (uint64_t)interlocked_or_64((volatile int64_t*)&destination, (int64_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(uint64_t, -1, destination);
```

Comment:
>uint64_t, -1 [](start = 21, length = 12)

We should not use -1 as expected, we should use an unsigned value

Reply:
Here and above in 738

---
In reply to: [427765880](https://github.com/Azure/azure-c-util/pull/43#discussion_r427765880) [](ancestors = 427765880)

Reply:
And in the other places too

---
In reply to: [427765967](https://github.com/Azure/azure-c-util/pull/43#discussion_r427765967) [](ancestors = 427765967,427765880)